### PR TITLE
Typescript support

### PR DESCRIPTION
**Pull request by lfades**
The following types are very basic but is enough, the `data` in a graphql response is typed as `object`.

### CODE REVIEW

File: src/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+declare module 'graphql-crunch' {
+  export function crunch(data: object | object[]): any[]
```
Comment (kachkaev) [inline src/index.d.ts]: Can’t data be just `any`? I’d also say the result is an array of (string, number, boolean, array of numbers or object with string keys and numeric values)

File: src/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+
+  export function uncrunch(data: any[]): object | object[]
```
Comment (kachkaev) [inline src/index.d.ts]: I’m not sure that `object | object[]` is the best output format here.  Accessing `result.something` will cause a ts exception, at least in `noImplicitAny` mode (hope I remember this correctly as i’m from the phone).

File: src/index.d.ts (hunk 1 above)
Comment (lfades) [reply src/index.d.ts]: I wasn't sure to know all the possible options for the result, but now yeah, we can definitely remove `any` from the result and replace `object` with `any`

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+declare module 'graphql-crunch' {
+  export type Uncrunched =
+    | { [key: string]: any }
+    | Array<{ [key: string]: any }>
```
Comment (kachkaev) [inline src/index.d.ts]: Types in TypeScript are usually nouns or nouns with adjectives. How about using `CrunchedData`/`UncrunchedData` instead of `Crunched`/`Uncrunched`? I also suspect that the algorithm can work well if un-crunched data is a scalar (integer/string/boolean), null or even undefined. In this case, we probably don't need `UncrunchedData` at all and can simply use `any`, just [like `JSON.stringify()`](https://github.com/Microsoft/TypeScript/blob/deef6cd88cf5d7f84146373fcccf86ada79c3aa3/lib/lib.es5.d.ts#L988).

@jamesreggio am I right that `crunch` will work for the same range of values as `JSON.stringify()`? It'd be great to see various edge cases in tests to avoid this question in future 🙂

Comment (lfades) [reply src/index.d.ts]: @stevekrenzel don't worry, I appreciate all of this a lot

I'll use `any` instead of the current type and rename `Crunched` with `CrunchedData`, btw `JSON.parse` returns `any` but I prefer to use `CrunchedData` because the result is always an array and it currently supports all the examples in the tests

**Comment by stevekrenzel** [comment]
🎉 Thanks!

**Comment by jamesreggio** [comment]
Sorry for jumping in post-merge, but I have some questions that I think y'all might be able to answer. (I'm a TypeScript n00b, so forgive my ignorance.)

1. It feels wrong to me to specify a detailed type union for `CrunchedData`. `CrunchedData` should _always_ be treated as opaque, and if the type system can help to enforce that constraint, I think it should. Is there any value in having this detailed union type for `CrunchedData` right now? And if we agreed to make it opaque, would we just make `CrunchedData` an alias for `any`?

2. After having witnessed a lot of type-system hand-wringing in Apollo Client, I expect that people will complain about the `any` type going in and out of the `crunch` and `uncrunch` methods. I think a better way to solve this would be to accept a template type of `TData`, and to use it in lieu of `any`. Any thoughts about or opposition to this?

**Comment by lfades** [comment]
For the first point I don't know if we can actually have opaque values in typescript, I did a little research and according to this [post](https://codemix.com/opaque-types-in-javascript/) and some github issues types are `transparent` by default, so for example if `CrunchedData` is `any` then any other type that is also `any` will be a valid matching peer for `CrunchedData` and in the case of my editor (VSCode) I'll see `any` instead of `CrunchedData`, keeping the above in mind I don't think that changing the current interface for `any` is an improvement.

Now for the second point I think a param that can replace the current `any` would be useful because `uncrunched` will return the same data received by `crunch`, the downside is that those are two separated methods that don't rely in each other, so how do you know which type to use ? of course `uncrunched` can also receive a type and return that same type, but it would be just the same as using `as`
